File: prod/butterfly.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#include <memory>
#include <boost/math/distributions/normal.hpp>

// Define the OptionPricingStrategy interface
class OptionPricingStrategy {
public:
    virtual double calculate_option_price(double S, double K, double T, double r, double sigma) const = 0;
    virtual ~OptionPricingStrategy() = default;  // Add a virtual destructor
};

// Implement the Black-Scholes pricing strategy
class BlackScholesPricing : public OptionPricingStrategy {
public:
    double calculate_option_price(double S, double K, double T, double r, double sigma) const override {
        using namespace boost::math;

        double d1 = (1 / (sigma * sqrt(T))) * (log(S / K) + (r + (sigma * sigma) / 2) * T);
        double d2 = d1 - sigma * sqrt(T);
        double C = S * cdf(normal(), d1) - K * exp(-r * T) * cdf(normal(), d2);
        return C;
    }
};

// Implement a factory for creating pricing strategies
class PricingStrategyFactory {
public:
    static std::unique_ptr<OptionPricingStrategy> create_strategy(const std::string& strategy_type) {
        if (strategy_type == "BlackScholes") {
            return std::make_unique<BlackScholesPricing>();
        }
        // Add more strategies here as needed
        throw std::invalid_argument("Invalid strategy type");
    }
};

// Context class that uses the selected pricing strategy
class OptionPricingContext {
private:
    std::unique_ptr<OptionPricingStrategy> pricing_strategy;

public:
    OptionPricingContext(std::unique_ptr<OptionPricingStrategy> strategy) : pricing_strategy(std::move(strategy)) {}

    double calculate_option_price(double S, double K, double T, double r, double sigma) const {
        return pricing_strategy->calculate_option_price(S, K, T, r, sigma);
    }
};

int main() {
    // Butterfly spread parameters
    double S = 100;       // Current stock price
    double K1 = 95;       // Lower strike price
    double K2 = 100;      // Middle strike price
    double K3 = 105;      // Higher strike price
    double T = 1;         // Time to expiration (in years)
    double r = 0.05;      // Risk-free interest rate
    double sigma = 0.2;   // Volatility

    // Create a context with the selected pricing strategy
    std::string strategy_type = "BlackScholes";  // You can change this to select a different strategy
    auto pricing_strategy = PricingStrategyFactory::create_strategy(strategy_type);
    OptionPricingContext pricing_context(std::move(pricing_strategy));

    // Calculate the individual call option prices
    double C1 = pricing_context.calculate_option_price(S, K1, T, r, sigma);
    double C2 = pricing_context.calculate_option_price(S, K2, T, r, sigma);
    double C3 = pricing_context.calculate_option_price(S, K3, T, r, sigma);

    // Calculate the net cost of the long call butterfly spread
    double cost = C1 - 2 * C2 + C3;

    std::cout << "Cost of Long Call Butterfly Spread using " << strategy_type << ": " << cost << std::endl;

    return 0;
}

File: prototype/butterfly.py
import math
from abc import ABC, abstractmethod
import scipy.stats as stats


# Define the OptionPricingStrategy interface
class OptionPricingStrategy(ABC):
    @abstractmethod
    def calculate_option_price(self, S, K, T, r, sigma):
        pass


# Implement the Black-Scholes pricing strategy
class BlackScholesPricing(OptionPricingStrategy):
    def calculate_option_price(self, S, K, T, r, sigma):
        d1 = (1 / (sigma * (T ** 0.5))) * (math.log(S / K) + (r + (sigma ** 2) / 2) * T)
        d2 = d1 - sigma * (T ** 0.5)
        C = S * stats.norm.cdf(d1) - K * math.exp(-r * T) * stats.norm.cdf(d2)
        return C


# Implement a factory for creating pricing strategies
class PricingStrategyFactory:
    @staticmethod
    def create_strategy(strategy_type):
        if strategy_type == "BlackScholes":
            return BlackScholesPricing()
        # Add more strategies here as needed
        raise ValueError("Invalid strategy type")


# Context class that uses the selected pricing strategy
class OptionPricingContext:
    def __init__(self, pricing_strategy):
        self.pricing_strategy = pricing_strategy

    def calculate_option_price(self, S, K, T, r, sigma):
        return self.pricing_strategy.calculate_option_price(S, K, T, r, sigma)


# Butterfly spread parameters
S = 100  # Current stock price
K1 = 95  # Lower strike price
K2 = 100  # Middle strike price
K3 = 105  # Higher strike price
T = 1  # Time to expiration (in years)
r = 0.05  # Risk-free interest rate
sigma = 0.2  # Volatility

# Create a context with the selected pricing strategy
strategy_type = "BlackScholes"  # You can change this to select a different strategy
pricing_strategy = PricingStrategyFactory.create_strategy(strategy_type)
pricing_context = OptionPricingContext(pricing_strategy)

# Calculate the individual call option prices
C1 = pricing_context.calculate_option_price(S, K1, T, r, sigma)
C2 = pricing_context.calculate_option_price(S, K2, T, r, sigma)
C3 = pricing_context.calculate_option_price(S, K3, T, r, sigma)

# Calculate the net cost of the long call butterfly spread
cost = C1 - 2 * C2 + C3

print(f"Cost of Long Call Butterfly Spread using {strategy_type}: {cost:.2f}")
